### DISCUSSION TITLE
vu meters.

### DISCUSSION THREAD

**Discussion by parabyte** [pschatzmann/arduino-snapclient#35] [Q&A] [answered]
okay im terribly not good at coding, but do enjoy the challenge but i seem to of hit a brick wall knowledge wise, and desperately seek wisdom.

on the audiotools page, there is projects featured and there is this wireless vu meter, I am essentially using the snapclient i2s example with very minimal changes, with this from one of your featured projects:

class MyI2S : public I2SStream {
  public:
    size_t write(const uint8_t *data, size_t len) override {
      int16_t *buffer16 = (int16_t*)data;
      int samples16 = len / 2;
      right = 0;
      left = 0;
      int16_t lmax = -32767;
      int16_t lmin = 32767;
      int16_t rmax = -32767;
      int16_t rmin = 32767;
      float factor = samples16 / 2;
      for (int j = 0; j < samples16; j = j + 2) {
        if (buffer16[j] < lmin) lmin = buffer16[j]; 
        if (buffer16[j] > lmax) lmax = buffer16[j];
        if (buffer16[j + 1] < rmin) rmin = buffer16[j + 1]; 
        if (buffer16[j + 1] > rmax) rmax = buffer16[j + 1];
      }
      left = abs(lmax - lmin);
      right = abs(rmax - rmin);
      sampled = true;
      return I2SStream::write(data, len);
    }
};

I cannot seem get this code to run real time, snapclient seems to be using up all available resources. 

i am using multi and adding this as one of the streams i believe you call them.

many thanks in advance,

**Comment by pschatzmann** [comment, marked as answer]
Why don't you just use the [existing VolumeMeter](https://github.com/pschatzmann/arduino-audio-tools/wiki/Determining-the-Volume) ?

What do you mean with "snapclient seems to be using up all available resources" ?

ps. you messed up with the data type for buffer16: it should be a pointer!

**Reply by parabyte** [reply]
once i fixed the data type in my code, stuff started to work nearly correctly.
but why reinvent the wheel when you already done the same thing, so i will use what you have suggested, thanks :)
